test/controllers/api_users/permissions_controller_test.rb

```diff
@@ -194,24 +194,42 @@ class ApiUsers::PermissionsControllerTest < ActionController::TestCase
       assert_same_elements [other_permission, not_from_ui_permission], api_user.supported_permissions
     end
 
-    should "prevent permissions being added for other apps" do
-      other_application = create(:application, with_non_delegatable_supported_permissions: %w[other])
-      application = create(:application)
-      api_user = create(:api_user)
-      create(:access_token, application:, resource_owner_id: api_user.id)
+    should "when updating permissions for app A, prevent additionally adding or removing permissions for app B" do
+      application_a = create(:application, with_non_delegatable_supported_permissions: %w[other])
+      application_a_old_permission = create(:supported_permission, application: application_a)
+      application_a_new_permission = create(:supported_permission, application: application_a)
+
+      application_b = create(:application)
+      application_b_old_permission = create(:supported_permission, application: application_b)
+      application_b_new_permission = create(:supported_permission, application: application_b)
+
+      api_user = create(:api_user,
+                        with_signin_permissions_for: [application_a, application_b],
+                        with_permissions: {
+                          application_a => [application_a_old_permission.name],
+                          application_b => [application_b_old_permission.name],
+                        })
+
+      create(:access_token, application: application_a, resource_owner_id: api_user.id)
```

Comment:
I was going to ask if we need an access token for `application_b`, but then thought that it's only used when setting the application from `params[:application_id]` in the controller. But I can't see that the resource owner matters in that context, so it wouldn't matter whether the actual API user has an access token. Is there somewhere else that the resource owner becomes relevant in this controller/its dependencies?

Reply:
We check that the user has a token that matches the params of the `application_id` we're trying to update [here](https://github.com/alphagov/signon/blob/main/app/controllers/api_users/permissions_controller.rb#L37) - `@api_user` there is the resource owner.